numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -110,42 +130,77 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
         return NPY_SUCCEED;
     }
 
-    len = PySequence_Size(obj);
-    if (len == -1) {
-        /* Check to see if it is an integer number */
-        if (PyNumber_Check(obj)) {
-            /*
-             * After the deprecation the PyNumber_Check could be replaced
-             * by PyIndex_Check.
-             * FIXME 1.9 ?
-             */
-            len = 1;
+    PyObject *seq_obj = NULL;
+
+    /*
+    * If obj is a scalar we skip all the useless computations and jump to
+    * intp_from_scalar as soon as possible.
+    */
+    if (!PyLong_CheckExact(obj) && PySequence_Check(obj)) {
+        seq_obj = PySequence_Fast(obj,
+               "Expected an integer or an iterable/sequence of integers.");
+        if (seq_obj == NULL) {
+            // we ignore this error since it is managed below
+            PyErr_Clear();
```

Comment:
Ah, we should be able to cover this, but it is a bit tricky, something like:
```
class BadSequenceButInt:
    def __len__(self):
        return 4

    def __getitem__(self, item):
        raise ValueError("Bad sequence!")

    def __index__(self):
         return 2
```
I guess we change the behaviour of such a ridiculous object (seems it was previously rejected).  But I think that is good (or fine).

Reply:
Should I ignore this one?

Reply:
Yes, fine to ignore it, I will add that `np.array(0)` test that I mentioned.